File: solutions/javascript/grade-school/1/grade-school.js
//
// This is only a SKELETON file for the 'Grade School' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class GradeSchool {
  constructor () {
    this.grades = {}
  }

  add (name, number) {
    console.log('add',Object.values(this.grades))
    let doesExist = Object.values(this.grades).some((students) => {
      return students.includes(name)
    })

    if (doesExist) return false

    if (!this.grades[number]) {
      this.grades[number] = []
    }
    
      this.grades[number].push(name)
      this.grades[number].sort()
      return true
  }
  
  roster() {
    return Object.values(this.grades).flat()
  }

  grade(number) {
    if (!this.grades[number]) {
      return []
    }
    return this.grades[number]
  }
}

File: solutions/cpp/grade-school/1/grade_school.cpp
#include <algorithm>
#include "grade_school.h"

namespace grade_school {

    void school::add(const std::string& name, unsigned grade) {
        auto pos = _roster.find(grade);
        if (pos != _roster.end()) {
            auto& v = pos->second;
            auto new_pos = std::lower_bound(v.begin(), v.end(), name);
            v.insert(new_pos, name);
        } else {
            _roster[grade] = {name};
        }
    }

    std::map<int, std::vector<std::string>> school::roster() const {
        return _roster;
    }

    std::vector<std::string> school::grade(unsigned grade) const {
        auto pos = _roster.find(grade);
        if (pos != _roster.end()) {
            return pos->second;
        } else {
            return {};
        }
    }

}  // namespace grade_school